### PR TITLE
Quesada chess jog 3

### PR DESCRIPTION
**Pull request by dianyleyva**
- I have a lot of queries (because I like them and wanted to try them lol). But I'll take some time to see if I can narrow it down.
- Also, I still have the other PR open since I need to work on time complexity (nested loops) so I assume I can have have two PRs.
-Thank you so much for all your help :)

### CODE REVIEW

**Comment by dianyleyva** [comment]
@robertlandis I will send you an email with this. But I'm attaching it here too.
Many of the feedback I got from PR #2 included restructuring/cleaning my code (change nested loops, change arrays to indexed by PK, downsize the amount of queries by doing more calc on the fly in php and less in sql). I applied them, but instead of doing it in PR #2 I did it over PR #3. I did so because I made so many changes in between them that I thought it would be just easier. 
But, what I submitted in PR #3 is way different to what I have now because of all those changes. So, I committed all the updates over PR#3. I just don't want you to spend time reviewing what I had before since I already changed it:). 
**Bottom line:**
I committed in PR#3 feedback from PR#2 and it should be ready to be reviewed.
I'm sorry if this makes things confusing. I hope it does not.

File: include/database_functions.php [hunk 1]
```diff
@@ -205,15 +151,58 @@ function insertClass($type, $link, $classDate, $studentId){
 ");
 }
 
+//If there is a better way to handle this please let me know. So, I have a credits table (Every row in that table in actually a purchase the student made).If
```
Comment (robertlandis) [inline include/database_functions.php:162]: I think this is alright, but another option would just be to add another credit to a student's account. Having a refund table means you could separate the two types of transactions, but it does add some complexity to the code. If that's worth it to you for the sake of having them be separate, then I say leave it; otherwise, I'd just use the existing credits table for this.

File: include/database_functions.php [hunk 2]
```diff
@@ -205,15 +151,58 @@ function insertClass($type, $link, $classDate, $studentId){
...
-function deleteStudents_TableDB($studentId){    
+function deleteStudent($studentId){    
     dbQuery("
         DELETE FROM students WHERE StudentId = $studentId
 ");
 }
 
+function deleteClass($classId){    
+    dbQuery("
+        DELETE FROM classes WHERE ClassId = $classId
```
Comment (robertlandis) [inline include/database_functions.php]: I can't remember if we discussed this, but deleting data from the DB can be a bit scary, since that means you'll never get it back (unless you have backups of some kind). To get around this, it's usually best practice to archive the data in the DB by setting a field for `DateArchived`. Then, if the value is NULL, you know it's an active class/student/etc., but if it has a date stored, then you know that they've been archived and when. Happy to discuss this more if that doesn't make sense.

File: include/form_functions.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+//Add student form   
+//here I use this form for both creating and editing a student. So, when creating a student
+//I don't pass the id and when editing I pass the id and thats how I diff it here. Id the studentId was passed 
+//Then I assign the values to the hidden fields in the form
+// --------------------------------------------------------------------------
+function openAddStudentForm(studentId){
+
+    if(studentId){
+        const student = document.getElementById('hiddenStudent').value;
```
Comment (robertlandis) [inline include/form_functions.js]: This isn't a wrong way to do this, per se, but it's not the typical pattern you'll see for handling this. Usually, you'll make an AJAX request to get the data from a PHP endpoint (which will return a JSON-encoded object), which you can then use to populate the form like you have below. Again, it's not wrong to do this, but I think it has more potential for bugs by storing a full JSON encoded string in a field, and also means you're loading data you may not need (e.g. if the student doesn't need to be edited, you're including this HTML somewhere on the page unnecessarily).

File: include/form_functions.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    
+    if(buttonclicked == 'Subtract'){                                                 //What's happening here is that if subtract was clicked I want the max to be set to remaining credits amount
+        document.getElementById('maxCredit').max = credits;                          //so the user cannot subtract more credits that the remaining ones. So if a user has 2 credits left a refund of more than
+        document.getElementById('maxCredit').placeholder = '1 - ' + credits;         //2 credits cannot be done.This ensure that only students with remaining credits can get a refund
+    }  
+    
+    if(buttonclicked == 'Add'){                                                 //Trying out I see that id I click subtract first and then add the placeholder and max values
+        document.getElementById('maxCredit').max = '100';                            //do not go back to normal and stay with the values I set above so I'm setting them here
+        document.getElementById('maxCredit').placeholder = '1 - 100';
+    }
```
Comment (robertlandis) [inline include/form_functions.js:45]: Where does the DB get updated here? This is another place where you might make an AJAX request to handle this.

File: admin/student_profile.php [hunk 5]
```diff
@@ -3,42 +3,108 @@
...
+
+      $date = formatDate($_REQUEST['classDate']." ".$_REQUEST['classTime'], 'Y/m/d H:i:s');
+      insertClass($_REQUEST['ctype'], $_REQUEST['czoomLink'], $date, $_REQUEST['studentId']);                           
+      header("location:? studentId={$_REQUEST['studentId']}");  
+      exit();                           
+    } 
+
+    if(isset($_REQUEST['studentDeleted'])){
+     
+      deleteStudent($_REQUEST['stdId']);
```
Comment (robertlandis) [inline admin/student_profile.php]: This is a bit nitpicky, but I would recommend keeping the way you reference ids consistent in your code. If this for `studentId`, I'd spell it out fully. This makes your code a little more readable for other devs, but also will help you remember what it's actually referencing when you look at the code in the future.

File: include/index_components.php [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                        ";       
+                                
+                    $currentTop+=$topPosition;
+                } 
+
+        echo"</div>";  
+}
+
+// --------------------------------------------------------------------------
+//I'm sure this is not what you had in mind haha since you were talking about a function to pass
```
Comment (robertlandis) [inline include/index_components.php:37]: I think this looks great!

File: include/list_students_components.php [hunk 7]
```diff
@@ -1,19 +1,18 @@
 <?php
 
+// --------------------------------------------------------------------------
+//I used an emoji for search bar icon lol. Is this ok?
```
Comment (robertlandis) [inline include/list_students_components.php:4]: I think that's fine, yeah!

**Review by robertlandis — CHANGES_REQUESTED** [review]

File: include/database_functions.php (hunk 1 above)
Comment (dianyleyva) [reply include/database_functions.php:162]: After we discussed this during Jog 3, I talked to Yuni, and we discussed that he would allow students to book a class even if they did not pay for it unless it is a new student. But if a student has taken classes with him during a certain time, they could book a class and pay for it afterward. I will have to add the ability to block/allow booking classes without credits. So, I thought that if we do so, it could be good to leave the refund table, so we know that negative entries in the credits table is a booked class that was not paid yet. What do you think about that?

File: include/database_functions.php (hunk 2 above)
Comment (dianyleyva) [reply include/database_functions.php]: Sounds great! I will do this for students and classes. What I'm thinking is that since credits do not get deleted, but refunded, then I won't put it on the credits table.

Comment (dianyleyva) [reply include/database_functions.php]: When doing this I added a where clause (WHERE DateArchived is NULL) to retrieve data. I did not do it in all queries here because I got rid of most of them in Jog 4. So I only added it to the ones I currently have in the current state on my code.

File: include/form_functions.js (hunk 3 above)
Comment (dianyleyva) [reply include/form_functions.js]: Ohhh! I was sure hiding things around wasn't the way to go around this lol. Let me look into "AJAX request" now.

File: include/form_functions.js (hunk 4 above)
Comment (dianyleyva) [reply include/form_functions.js:45]: So here, the add/subtract buttons open a form but instead of loading the current credits to be edited in the form, what I'm doing is that I show a type number field where I restrict the max number and placeholder depending on the remaining credits the student has. The DB is updated when the form is submitted, if it was subtracted the change will go to the refund table, and if added will go to the credits table. 
I'm trying to see how to use AJAX request here. 
To get the credits through AJAX instead of passing it to the function as parameter? 
Or maybe I can have + and - buttons that do not open a form but that add/subtract credits using AJAX and the credits amount is updated it without reloading the page?

Comment (robertlandis) [reply include/form_functions.js:45]: Ah okay, I didn't realize it opened a form. That's totally fine if you prefer that UX, I just wanted to make sure there was something that communicated with the DB and it wasn't a front-end only update that was happening.

File: include/database_functions.php (hunk 2 above)
Comment (robertlandis) [reply include/database_functions.php]: Sounds good to me!

**Comment by dianyleyva** [comment]
@robertlandis I just committed the changes after applying feedback. This is just fyi, I know you must have had a rough day yesterday with the DDoS, so no worries and no hurry. Thanks a lot

File: include/form_functions.js [hunk 8]
```diff
@@ -31,8 +32,18 @@ function closeAddStudentForm(){
 //Add credits form   
 // --------------------------------------------------------------------------
 function openCreditForm(buttonclicked, credits){ 
-    document.getElementById('hiddenValue').value = buttonclicked;                             //I'm setting the id parameter to the type hidden field so I can know wich button was clicked in the Request
+    document.getElementById('hiddenButtonName').value = buttonclicked;                             //I'm setting the id parameter to the type hidden field so I can know wich button was clicked in the Request
     
+    //AJAX request to get the student from the DB
```
Comment (dianyleyva) [inline include/form_functions.js]: I put this here because I was thinking about doing AJAX request instead of passing the credits, (so I was about to work with this code as template) but then I realized that the remaining credits are being displayed in the page anyway so I just keep it like this and then I deleted this code (You will see it in the following commit). Sorry for any confusion. I didn't realized I committed with this here

File: include/AJAX_Requests.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
+<?php
+include('../include/initialize.php'); 
+
+//I'm planning to send all my AJAX requests to this file from now as I get more familiar with the concept. Is this ok?
```
Comment (robertlandis) [inline include/AJAX_Requests.php]: This is alright, but the way we approach it is to have an `ajax` directory, and put a different file for each request. So you might have something like:
- `ajax/get_student.php`
- `ajax/get_class.php`
- `ajax/login.php`

Since the whole file is going to load, you usually just want to load the code you need for the specific AJAX request you're making. It's not a huge deal to have it this way, but I think you can probably make it a little more organized to have separate files vs. one larger one.

File: include/AJAX_Requests.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+if(isset($_GET['StudentIdToEdit'])){
+    $studentId = $_GET['StudentIdToEdit'];
+
+    echo json_encode(getOneStudent($studentId));
+} 
+
+if(isset($_POST['SavePrivateNotes'])){
+
+    $commments = json_decode($_POST['SavePrivateNotes']);
```
Comment (robertlandis) [inline include/AJAX_Requests.php]: You don't have to change this, but if you'd rather work with a PHP array instead of a generic PHP object, you can pass `true` to the second parameter of `json_decode` and it'll create an array instead. It's mostly a preference thing, but I tend to prefer using an array over an object personally.

**Review by robertlandis — CHANGES_REQUESTED** [review]

File: include/AJAX_Requests.php (hunk 10 above)
Comment (dianyleyva) [reply include/AJAX_Requests.php]: Great! I also preferer arrays.

**Comment by dianyleyva** [comment]
I did the last changes you recommended!

**Review by robertlandis — APPROVED** [review]

**Comment by robertlandis** [comment]
@Diany-Leyva This looks great! I'm going to merge it in to `main` now. When you get a chance, will you merge `main` into the other branch for the next PR review?
